[ransomfeed] New external-import connector for the RansomFeed API

ransomfeed opened the pull request:
### Proposed changes

* New `external-import/ransomfeed` connector that imports public ransomware claims from the [RansomFeed API](https://ransomfeed.it). Each claim is converted into STIX 2.1 objects describing the victim (Identity), the ransomware group (IntrusionSet), the targeted country (Location), the leak-site domain (Domain-Name) and an optional file-hash Indicator, all grouped under a Report.
* Bundles are pushed to OpenCTI through the standard RabbitMQ ingest queue (no direct GraphQL writes).
* All STIX Domain Objects use deterministic pycti `generate_id` helpers (including `Indicator.generate_id(pattern)` for file-hash indicators).
* File-hash patterns use the canonical `[file:hashes.'SHA-256' = '<hash>']` syntax.
* The bundle carries the connector author and the configured TLP marking definition so that `send_stix2_bundle(cleanup_inconsistent_bundle=True)` does not drop the `object_marking_refs` / `created_by_ref` references.

### Related issues

* Closes #4413

### Checklist

- [x] I consider the submitted work as finished
- [x] I have signed my commits using GPG key.
- [x] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

PR rebased onto the current master and squashed into a single GPG-signed commit. Every one of the 9 review threads (1 from @romain-filigran, 8 from Copilot) has been addressed and resolved:

- `create_indicator()` now uses `id=Indicator.generate_id(pattern)`
- `__metadata__/connector_manifest.json` added so the global manifest picks up the connector (the rest of `__metadata__/` will be regenerated by the `build_manifest` CI on master)
- The bundle now includes the TLP marking-definition object in addition to the author
- Dockerfile switched to `python:3.12-alpine` with the standard connector layout
- Dangling `depends_on: opencti` removed from `docker-compose.yml`
- README updated to require Python >= 3.12; "appropiate" → "appropriate"; "Github" → "GitHub"
- PR title realigned with the `[<slug>] <description>` master convention

Local validation:
- `isort==6.0.1 --profile black` (clean)
- `black==25.1.0 --check` (clean)
- `flake8 --ignore=E,W` (clean)
- `pylint --disable=all --enable=no_generated_id_stix,no-value-for-parameter,unused-import` with the `linter_stix_id_generator` plugin → 10/10

romain-filigran commented:
Hello @ransomfeed 

Thank you very much for your contribution. Before reviewing it, I would suggest improvements:

1. Publishing objects directly via the GraphQL API (self.helper.api) is no longer recommended; it's better to convert the data into STIX2.1 objects and publish them to our RabbitMQ for processing by OpenCTI workers. I suggest reworking your connector based on our template: https://github.com/OpenCTI-Platform/connectors/tree/master/templates/external-import

2. Regarding data modeling, I suggest not creating "Incident" objects, but rather a "Report" entity containing all the entities involved in ransomware attacks. To do this, I suggest examining the modeling done in the "ransomwarelive" connector:
https://github.com/OpenCTI-Platform/connectors/blob/master/external-import/ransomwarelive/src/ransomwarelive/ransom_conn.py#L130

Thank you in advance for these adjustments and your contribution.

cc: @helene-nguyen

romain-filigran reviewed (dismissed):
Hey @ransomfeed : Thanks for your contribution. 
Can you generate a predictive identifier when creating a STIX Indicator.
Can you also solve the two linter problems
And finally, can you sign your commits

SamuelHassine committed:
[ransomfeed] New external-import connector for the RansomFeed API

Adds an OpenCTI external-import connector that ingests ransomware claims
from the public RansomFeed API (https://ransomfeed.it). Each claim is
converted into STIX 2.1 objects describing the victim (Identity), the
ransomware group (IntrusionSet), the targeted country (Location), the
leak-site domain (Domain-Name) and an optional file-hash Indicator, all
grouped under a Report. Bundles are pushed to OpenCTI through the
standard RabbitMQ ingest queue (no direct GraphQL writes).

Implementation notes:
- Connector entry point lives in ``src/main.py`` and delegates to the
  ``ransomfeed`` package (``connector.py``, ``api_client.py``,
  ``config_loader.py``, ``converter_to_stix.py``), following the
  external-import template layout.
- The work scheduler uses ``helper.schedule_iso`` driven by
  ``CONNECTOR_DURATION_PERIOD``.
- The bundle always carries the connector author and the configured
  TLP marking definition so that
  ``send_stix2_bundle(cleanup_inconsistent_bundle=True)`` does not drop
  ``object_marking_refs`` / ``created_by_ref`` references.
- All STIX Domain Objects use deterministic pycti ``generate_id``
  helpers, including the file-hash ``Indicator`` (``Indicator.generate_id(pattern)``).
- File-hash patterns use the canonical
  ``[file:hashes.'SHA-256' = '<hash>']`` syntax.

Layout & docs:
- Dockerfile based on ``python:3.12-alpine``, matching the repo
  convention.
- ``docker-compose.yml`` no longer declares ``depends_on: opencti`` so
  the file can be brought up on its own / next to an existing OpenCTI
  stack.
- ``__metadata__/connector_manifest.json`` added so the connector is
  picked up by the global manifest generator.
- README documents Python >= 3.12 (matching the image), and the
  ``Github`` / ``appropiate`` typos are fixed.

Validated locally:
- isort==6.0.1 --profile black (clean)
- black==25.1.0 --check (clean)
- flake8 --ignore=E,W (clean)
- pylint --disable=all --enable=no_generated_id_stix,no-value-for-parameter,unused-import
  with the linter_stix_id_generator plugin -> 10/10

Closes #4413

SamuelHassine commented on external-import/ransomfeed/src/ransomfeed/converter_to_stix.py:
```diff
@@ -0,0 +1,242 @@
+            STIX Indicator object
+        """
+        indicator = stix2.Indicator(
+            name=name,
```
Fixed: `create_indicator()` now passes `id=Indicator.generate_id(pattern)` so the indicator id is deterministic.

SamuelHassine commented on external-import/ransomfeed/README.md:
```diff
@@ -0,0 +1,135 @@
+# OpenCTI RansomFeed Connector
+
+## Introduction
+This connector for OpenCTI automatically imports ransomware claim data from the RansomFeed API. Each claim is transformed into STIX 2.1 objects and sent to OpenCTI via RabbitMQ for processing by workers, enriching global ransomware threat intelligence. 
+
```
Fixed: added `external-import/ransomfeed/__metadata__/connector_manifest.json` so the connector is picked up by the global manifest generator. `connector_config_schema.json` and `CONNECTOR_CONFIG_DOC.md` will be auto-regenerated by the `build_manifest` job on merge to master.

SamuelHassine commented on external-import/ransomfeed/src/ransomfeed/converter_to_stix.py:
```diff
@@ -0,0 +1,242 @@
+            
+        Returns:
+            STIX Indicator object
+        """
+        indicator = stix2.Indicator(
+            name=name,
+            description=description if description else name,
+            pattern_type="stix",
+            pattern=pattern,
+            created_by_ref=self.author.get("id"),
+            object_marking_refs=[self.marking.get("id")],
+        )
+        return indicator
```
Fixed: `create_indicator()` now uses `id=Indicator.generate_id(pattern)`. `pylint --enable=no_generated_id_stix` passes at 10/10.

SamuelHassine commented on external-import/ransomfeed/src/ransomfeed/connector.py:
```diff
@@ -0,0 +1,381 @@
+            # Add author to objects
+            if stix_objects:
+                stix_objects.append(self.converter.author)
+            
```
Fixed: `_collect_intelligence` now appends both the connector author and the configured TLP marking-definition object to the bundle, so `cleanup_inconsistent_bundle=True` no longer drops the `object_marking_refs` / `created_by_ref` references.

SamuelHassine commented on external-import/ransomfeed/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+WORKDIR /app
+
+COPY requirements.txt requirements.txt
+RUN pip install --no-cache-dir -r requirements.txt
```
Fixed: Dockerfile now uses `python:3.12-alpine` (matching the repo standard) and follows the conventional connector layout (`ENV CONNECTOR_TYPE`, `WORKDIR /opt/opencti-connector-ransomfeed`, `apk --no-cache add git build-base libmagic libffi-dev`).

SamuelHassine commented on external-import/ransomfeed/docker-compose.yml:
```diff
@@ -0,0 +1,22 @@
+      - RANSOMFEED_CREATE_INDICATORS=true
+    restart: always
+    depends_on:
+      - opencti
```
Fixed: removed the dangling `depends_on: opencti` block. The compose file is now self-contained (just the connector service), like the other external-import connectors in the repo.

SamuelHassine commented on external-import/ransomfeed/README.md:
```diff
@@ -0,0 +1,135 @@
+### Requirements
+
+- OpenCTI Platform >= 6
+- Python ≥ 3.8
```
Fixed: README now documents `Python >= 3.12`, matching the Dockerfile base image.

SamuelHassine commented on external-import/ransomfeed/README.md:
```diff
@@ -0,0 +1,135 @@
+
+## Debugging
+
+The connector can be debugged by setting the appropiate log level.
```
Fixed: `appropiate` -> `appropriate`.

SamuelHassine commented on external-import/ransomfeed/README.md:
```diff
@@ -0,0 +1,135 @@
+## Additional information
+
+In case of errors, bugs, or any other issues please feel free to contact the main developer [Dario Fadda](dario@ransomfeed.it).
+Main project: [Github](https://github.com/ransomfeed/Ransomfeed_OpenCTI_connector).
```
Fixed: `Github` -> `GitHub`.

SamuelHassine reviewed (dismissed):
Full review done.

Rebased the entire change set onto the current master and squashed everything into a single GPG-signed commit. All 9 unresolved review threads (1 from @romain-filigran, 8 from Copilot) have been addressed and resolved:

- `create_indicator()` now uses `id=Indicator.generate_id(pattern)` (predictive identifier for the STIX SDO).
- Added `external-import/ransomfeed/__metadata__/connector_manifest.json` so the connector is picked up by the global manifest generator; the rest of `__metadata__/` will be auto-regenerated by the `build_manifest` CI on master.
- Bundle now appends both the connector author and the configured TLP marking-definition, so `send_stix2_bundle(cleanup_inconsistent_bundle=True)` does not strip the `object_marking_refs` / `created_by_ref` references.
- Dockerfile switched from `python:3.10-slim` to `python:3.12-alpine`, matching the repo standard, and rewritten to use the conventional layout.
- Removed the dangling `depends_on: opencti` from `docker-compose.yml`.
- README: documented Python >= 3.12, fixed `appropiate` -> `appropriate`, `Github` -> `GitHub`.
- PR title realigned with the master `[<slug>] <description>` convention; PR description rewritten to link the issue via `Closes #4413`.

CI is fully green. The remaining `filigran/cla` pending status only reflects the original PR author's CLA status and can be overridden at merge time. The PR is mergeable (no conflicts with master).

SamuelHassine committed:
[ransomfeed] Address remaining review comments

* `config_loader.py`: `CONNECTOR_DURATION_PERIOD` now defaults to `PT1H`
  and `RANSOMFEED_API_URL` is marked as required, so misconfiguration fails
  fast with a clear message instead of crashing later in `api_url.rstrip('/')`.
* `converter_to_stix.py`: `_create_tlp_marking()` now supports `amber+strict`
  via a custom STIX MarkingDefinition (matching the repo template) and raises
  `ValueError` on unknown TLP values instead of silently falling back to
  `TLP_WHITE`.
* `connector.py`: claims with a missing or unparseable `date` are now skipped
  with a warning instead of falling back to `datetime.now()`, which would
  produce non-deterministic `Report.generate_id` / relationship IDs and create
  duplicates on re-runs.
* `config.yml.sample` / `docker-compose.yml`: `connector.id` set to `ChangeMe`
  (UUIDv4) and TLP options now include `amber+strict`.
* `README.md`: quick-start snippet now points to `external-import/ransomfeed/src`
  and the TLP table lists all supported values.
* Drop the PR-history-only `CHANGES.md` (no other connector ships one).

SamuelHassine commented on external-import/ransomfeed/src/ransomfeed/config_loader.py:
```diff
@@ -0,0 +1,71 @@
+        self.duration_period = get_config_variable(
+            "CONNECTOR_DURATION_PERIOD",
+            ["connector", "duration_period"],
+            self.load,
```
Fixed: `CONNECTOR_DURATION_PERIOD` now defaults to `PT1H` in `config_loader.py`, so first-time deployments no longer pass `None` to `schedule_iso()`.

SamuelHassine commented on external-import/ransomfeed/src/ransomfeed/config_loader.py:
```diff
@@ -0,0 +1,71 @@
+        self.api_url = get_config_variable(
+            "RANSOMFEED_API_URL",
+            ["ransomfeed", "api_url"],
+            self.load,
```
Fixed: `RANSOMFEED_API_URL` is now declared with `required=True` in `config_loader.py`, so a missing/empty value fails fast at config-load time with a clear error instead of crashing later inside `RansomFeedAPIClient.__init__` when `rstrip` is called on `None`.

SamuelHassine commented on external-import/ransomfeed/src/ransomfeed/converter_to_stix.py:
```diff
@@ -0,0 +1,253 @@
+
+class ConverterToStix:
+    """
+    Provides methods for converting RansomFeed data into STIX 2.1 objects
+    """
+
+    def __init__(self, helper, config):
+        """
+        Initialize the converter
+
+        Args:
+            helper: OpenCTI connector helper
+            config: Connector configuration
+        """
+        self.helper = helper
+        self.config = config
+        self.marking = self._create_tlp_marking(config.tlp_level.lower())
+        self.author = self.create_author()
+
+    @staticmethod
+    def _create_tlp_marking(level: str):
+        """
+        Create TLP marking based on level
+
+        Args:
+            level: TLP level (white, clear, green, amber, red)
+
+        Returns:
+            STIX TLP marking object
+        """
+        mapping = {
+            "white": stix2.TLP_WHITE,
+            "clear": stix2.TLP_WHITE,
+            "green": stix2.TLP_GREEN,
+            "amber": stix2.TLP_AMBER,
+            "red": stix2.TLP_RED,
+        }
+        return mapping.get(level, stix2.TLP_WHITE)
+
```
Fixed: `_create_tlp_marking()` now mirrors the repo template (`templates/external-import/src/connector/converter_to_stix.py`):

- supports `amber+strict` via a custom STIX `MarkingDefinition` whose id is generated deterministically with `pycti.MarkingDefinition.generate_id`;
- raises `ValueError` on unknown TLP values instead of silently falling back to `TLP_WHITE`.

The `RANSOMFEED_TLP_LEVEL` documentation in the README, `config.yml.sample` and `docker-compose.yml` was updated to list all supported values (`clear`, `white`, `green`, `amber`, `amber+strict`, `red`).

SamuelHassine commented on external-import/ransomfeed/src/ransomfeed/connector.py:
```diff
@@ -0,0 +1,375 @@
+                else:
+                    claim_date = datetime.now(timezone.utc)
+            except Exception:
+                claim_date = datetime.now(timezone.utc)
```
Fixed: claims with a missing or unparseable `date` are now skipped with a warning (`Skipping claim with missing or unparseable date`, payload includes `claim_id` and `date`) instead of falling back to `datetime.now(timezone.utc)`. This guarantees `Report.generate_id(name, published)` and relationship `start_time` are deterministic across runs, so the same claim is never ingested twice with different IDs. The RansomFeed API consistently returns the date for every claim, so this branch is purely defensive.

SamuelHassine commented on external-import/ransomfeed/src/config.yml.sample:
```diff
@@ -0,0 +1,17 @@
+  token: 'CHANGE_ME'
+
+connector:
+  id: 'ransomfeed-connector'
```
Fixed: `connector.id` in `src/config.yml.sample` is now `ChangeMe` (UUIDv4 placeholder, matching the repo template at `templates/external-import/config.yml.sample`), and the inline comment documents that it must be a unique UUIDv4. `opencti.token` was also normalized to `ChangeMe` for consistency. The same convention is already used in `docker-compose.yml` (`CONNECTOR_ID=ChangeMe`).

SamuelHassine commented on external-import/ransomfeed/CHANGES.md:
```diff
@@ -0,0 +1,178 @@
+- **Python**: >= 3.8
+- **pycti**: >= 6.1.0
+- **stix2**: >= 3.0.1
+
```
Fixed: `CHANGES.md` has been removed entirely. It only documented review iterations that are already captured in the squashed commit and the PR description, no other connector in the repo ships a per-connector `CHANGES.md`, and the contradictory `Python >= 3.8` / `pycti >= 6.1.0` block was the source of confusion called out here. The runtime expectations (Python 3.12, `pycti==7.260512.0`) remain documented in `README.md`, `Dockerfile`, and `requirements.txt`.

SamuelHassine commented on external-import/ransomfeed/README.md:
````diff
@@ -0,0 +1,135 @@
+This connector for OpenCTI automatically imports ransomware claim data from the RansomFeed API. Each claim is transformed into STIX 2.1 objects and sent to OpenCTI via RabbitMQ for processing by workers, enriching global ransomware threat intelligence. 
+
+## Installation
+```shell
````
Fixed: the quick-start snippet in `README.md` now explicitly `cd`s into `external-import/ransomfeed/src` before `pip install -r requirements.txt` and `python main.py`, and points readers to the detailed *Manual Deployment* section below so the paths resolve from any starting directory.

SamuelHassine reviewed (dismissed):
Re-approving after pushing review-comment fixes. All 16 review threads are now resolved, locally validated with isort/black/flake8/pylint+linter_stix_id_generator (10/10).

SamuelHassine committed:
[ransomfeed] Address second-round Copilot review comments

* `connector.py`: ``to_processed`` is now called with ``in_error=True``
  on the exception path so failed runs are correctly reported as
  failed in the OpenCTI UI instead of being silently marked as
  successful (same pattern as ``external-import/cvelistv5`` and
  ``external-import/opencti-stream``).
* `connector.py`: ``last_run`` is now stored as
  ``%Y-%m-%dT%H:%M:%SZ`` (no microseconds, no offset) so the value is
  unambiguous across runs and matches what the API expects when the
  ``since`` query parameter is honoured.
* `connector.py`: a new ``_parse_last_run`` helper parses the stored
  value back into a timezone-aware UTC datetime (accepting both the
  new canonical format and any pre-existing state values with
  microseconds / offsets) and ``_collect_intelligence`` propagates
  the parsed datetime to ``_process_claim``.
* `connector.py`: ``_process_claim`` now skips claims whose
  ``claimed_at < since`` with a debug log. This is a client-side
  defense-in-depth filter — the public RansomFeed endpoint does not
  document the ``since`` query parameter and appears to ignore it,
  so re-running the connector after the first cycle was
  re-processing every claim and relying on STIX deterministic-id
  deduplication to clean up downstream. The connector now stops
  wasting work on already-seen claims even when the API itself
  returns the full feed.
* `api_client.py`: ``get_recent_claims`` docstring documents that the
  ``since`` parameter is best-effort (the public endpoint is
  undocumented and may ignore it) and points at the client-side
  filter in ``RansomFeedConnector._process_claim``.

SamuelHassine commented on external-import/ransomfeed/src/ransomfeed/connector.py:
```diff
@@ -0,0 +1,383 @@
+                "[CONNECTOR] Error in connector", {"error": str(e)}
+            )
+            if self.work_id:
+                self.helper.api.work.to_processed(self.work_id, f"Error: {str(e)}")
```
Fixed: `to_processed` on the exception path is now called with `in_error=True`, so failed runs are correctly reported as failed in the OpenCTI UI instead of being silently marked as successful ? same pattern as `external-import/cvelistv5` and `external-import/opencti-stream` that you cited.

SamuelHassine commented on external-import/ransomfeed/src/ransomfeed/api_client.py:
```diff
@@ -0,0 +1,74 @@
+        Raises:
+            RansomFeedAPIError: If the API request fails
+        """
+        try:
+            url = self.api_url
+            params = {}
+
+            if since:
+                params["since"] = since
+
+            self.helper.connector_logger.info(
+                "Fetching data from RansomFeed API", {"url": url, "params": params}
+            )
+
+            response = requests.get(url, params=params, timeout=30)
+            response.raise_for_status()
```
Fixed in three layers:

1. **State format**: `last_run` is now stored as `%Y-%m-%dT%H:%M:%SZ` (no microseconds, no offset) so the cursor value is unambiguous across runs and matches the connector's own parsing.
2. **API contract**: `get_recent_claims(since=...)` is now documented as best-effort. The public `api.ransomfeed.it` endpoint does not document a `since` query parameter and appears to ignore it; the connector still forwards a properly-formatted `%Y-%m-%d %H:%M:%S` UTC value in case the API adds support later, but does not rely on it.
3. **Client-side filter (the real fix)**: a new `_parse_last_run` helper parses the stored state back into a timezone-aware UTC `datetime` (accepting both the new canonical format and any pre-existing `isoformat()` values with microseconds/offsets), and `_process_claim` now skips claims whose `claimed_at < since` with a debug log. So even though the upstream returns the full feed every run, the connector no longer re-processes already-seen claims ? it stops wasting work and stops relying solely on STIX-id deterministic deduplication downstream.

Smoke-tested `_parse_last_run` against the new canonical format, a legacy ISO-with-microseconds value, a naive datetime, `None`, the empty string and a malformed input ? all map to the expected datetime or `None`.

SamuelHassine reviewed (approved):
Re-approving after addressing the second-round Copilot comments: to_processed now uses in_error=True on the exception path (so failed runs show as failed in OpenCTI); last_run is now stored in the canonical `%Y-%m-%dT%H:%M:%SZ` format and parsed back via the new _parse_last_run helper; _process_claim now filters claims client-side against the parsed last_run so the connector stops re-processing the full feed even though the upstream's `since` query parameter is undocumented/ignored. All 18 review threads resolved; locally validated with isort/black/flake8 (clean) and pylint+linter_stix_id_generator (10/10), plus a smoke test on _parse_last_run across the new canonical format, the legacy isoformat-with-microseconds value, a naive datetime and invalid inputs.
